### DISCUSSION TITLE
Translated all C++ files with AI. Cost? 1.15$ USD

### DISCUSSION THREAD

**Discussion by 0xeb** [OpenSalamander/salamander#50] [General]
I used GPT 3.5 Turbo APIs and custom private tooling to do the comments translation of all C++ files.
Header files next.

<img width="774" alt="image" src="https://github.com/OpenSalamander/salamander/assets/230851/3327bbc5-f9bd-404f-b8f8-fff8d1985a98">

Tokens:
<img width="427" alt="image" src="https://github.com/OpenSalamander/salamander/assets/230851/fcb4d86d-33c3-4543-9e2e-27308b44c14d">

Cost:
<img width="283" alt="image" src="https://github.com/OpenSalamander/salamander/assets/230851/e3196444-1959-4750-9497-db0626708955">

I think it will cost no more than 3$ to finish the whole thing.

I will review the changes and start making PRs.

**Comment by janrysavy** [comment]
Do you think it is possible to use already translated (reviewed) files to fine tune it? It would be probably better to use GPT-4?

**Reply by melloware** [reply]
I love this!  Working smarter not harder 😄

**Comment by janrysavy** [comment]
The translated files contain the following header line:
```// CommentsTranslationProject: TRANSLATED```

**Reply by 0xeb** [reply]
Indeed. I spotted that, and these files are skipped altogether.

<img width="684" alt="image" src="https://github.com/OpenSalamander/salamander/assets/230851/c481c2a1-628b-462f-b5ca-635f58df7545">

**Comment by 0xeb** [comment]
Why are many PRs still open from December?

- How much will it take to go over ~128 translated files?
- Prefer one PR per file?

**Reply by janrysavy** [reply]
Some files have already been reviewed and are waiting for fixes. I can take over and fix them myself. Others are waiting for a review, probably mine.

Is it possible to fine-tune GPT translation with already translated files (where we have Czech original and correct English translation)?

Why not use GPT-4 API instead of GPT-3.5?

I would prefer to start with only one PR per file and only with one or two files.

**Comment by 0xeb** [comment]
1. 3.5 Turbo for speed and cost considerations. I first tried to use local models (mixtral-dolphin), but they were too slow on my machine AND they were not good with Czech.

2. Yes, it is possible, but is it worthwhile the efforts?

I have attached a sample of the translation cache that will be used to apply the comments and that can be used for fine tuning:
[tcache.json](https://github.com/OpenSalamander/salamander/files/14839728/tcache.json) (just to give you an idea).

Fine tuning will also incur costs as well.

**Comment by janrysavy** [comment]
Do you have any idea how much would GPT-4 increase the price? Should I imagine something like order of x10? If we're talking in the order of tens of dollars, GPT-4 would make sense from my perspective for a higher quality translation.

**Reply by 0xeb** [reply]
You have to consider many factors going into the translation.

I spent the whole weekend tooling for this to automate it.

First, I use basic Google Translate API, but the translation was crappy.
Then local models supported by Ollama, not even better.
Finally, tried GPT 3.5 Turbo.


Now one of the BIGGEST factors in translation is not just the text, but the surrounding context. The code around, the function, and also how we take the comments. Translate one comment line by line? Try to group comments? How to tokenize, etc....

Not an easy problem given cost and computation constraints.

Let me make a PR of a few files and you can take a look. I think this automatic work got us 75% if not more to finish the translation task with automation, and 25% intervention and review.

If we find repeated bad translation, I can patch the translation table and re-apply on the source code.

**Reply by janrysavy** [reply]
Sounds great, let's give it a try...

**Comment by 0xeb** [comment]
Probably more than x10. Source: https://perplexity.ai/pro?referral_code=VYEVGH7M

Here is a side-by-side comparison of the costs for GPT-4 and GPT-3.5 based on the pricing information provided in the search results:

| Model                | Input Tokens (per 1K) | Output Tokens (per 1K) |
|----------------------|----------------------:|------------------------:|
| GPT-4 (8K context)   |                 $0.03 |                   $0.06 |
| GPT-4 (32K context)  |                 $0.06 |                   $0.12 |
| GPT-3.5 Turbo        |               $0.0015 |                  $0.002 |

To calculate the costs for 1,375,049 input tokens and 309,795 output tokens:

GPT-4 (8K context):
- Input tokens cost = 1,375,049 * ($0.03 / 1000) = $41.25
- Output tokens cost = 309,795 * ($0.06 / 1000) = $18.59
- Total cost = $41.25 + $18.59 = $59.84

GPT-4 (32K context):
- Input tokens cost = 1,375,049 * ($0.06 / 1000) = $82.50
- Output tokens cost = 309,795 * ($0.12 / 1000) = $37.18
- Total cost = $82.50 + $37.18 = $119.68

GPT-3.5 Turbo:
- Input tokens cost = 1,375,049 * ($0.0015 / 1000) = $2.06
- Output tokens cost = 309,795 * ($0.002 / 1000) = $0.62
- Total cost = $2.06 + $0.62 = $2.68

In summary, for the given token usage:
- GPT-4 (8K context) would cost $59.84
- GPT-4 (32K context) would cost $119.68 
- GPT-3.5 Turbo would cost only $2.68

The cost difference between GPT-4 and GPT-3.5 Turbo is substantial, with GPT-4 being over 22 times more expensive than GPT-3.5 Turbo for the 8K context model, and nearly 45 times more expensive for the 32K context model[1][2][11][17].

Citations:
[1] https://www.gptpromptshub.com/gpt-4-pricing-calculator
[2] https://invertedstone.com/calculators/openai-pricing/
[3] https://articlefiesta.com/seo-tools/token-calculator
[4] https://livechatai.com/gpt-4-turbo-pricing-calculator
[5] https://www.reddit.com/r/OpenAI/comments/12fz94t/anyone_know_how_to_calculate_the_api_cost_of_the/
[6] https://www.reddit.com/r/OpenAI/comments/13yuvzg/question_about_gpt_pricing_per_token/
[7] https://community.openai.com/t/estimating-gpt-api-conversation-costs-factoring-in-cumulative-input-and-output-tokens/408818
[8] https://community.openai.com/t/how-to-accurately-price-a-gpt-4-chatbot/347250
[9] https://docsbot.ai/tools/gpt-openai-api-pricing-calculator
[10] https://community.openai.com/t/how-much-are-the-cost-of-this-example-in-gpt-4/591747
[11] https://help.openai.com/en/articles/7127956-how-much-does-gpt-4-cost
[12] https://finetunedb.com/blog/how-much-does-it-cost-to-finetune-gpt-35/
[13] https://community.openai.com/t/gpt-4-pricing-for-chat-api/669700
[14] https://neoteric.eu/blog/how-much-does-it-cost-to-use-gpt-models-gpt-3-pricing-explained/
[15] https://www.reddit.com/r/OpenAI/comments/12r1zoa/has_anyone_compared_the_cost_of_using_gpt4_api_to/
[16] https://apidog.com/blog/openai-api-pricing/
[17] https://openai.com/pricing
[18] https://www.reddit.com/r/GPT3/comments/11tlg75/openai_is_expensive/
[19] https://gptforwork.com/tools/openai-chatgpt-api-pricing-calculator
[20] https://community.openai.com/t/gpt-3-5-turbo-api-pricing/612022

**Reply by janrysavy** [reply]
What parser do you use for extracting comments from source codes?

Have you considered the possibility of open-sourcing this project? I believe it would be useful for other projects as well.

**Reply by 0xeb** [reply]
I used libclang. For now, it is a bunch of scripts I put together over the weekend, not even close to being polished and only works for this project. First time, I do this kind of work.

But, because of it, there were some other work needed as well. Like source code patching, proper parsing flags from SLN/VCXPROJ files, translation research / various backend support, etc.

Maybe, I can clean up and open source eventually.

**Reply by janrysavy** [reply]
At Fine (primarily a Delphi shop) we use Clang AST directly for H2PAS translations. It works great.

Do you use the [Clang’s refactoring engine](https://clang.llvm.org/docs/RefactoringEngine.html#clang-s-refactoring-engine)?

**Reply by 0xeb** [reply]
Good question. I have not used Clang's refactoring engine. I wrote my own 'extent' base refactoring library. Once I have the needed Token or Node, I use its extent (start/end pos) to patch the text. Easier and faster for my purposes. My refactor script cannot handle real refactoring, but mostly code edits I would say.

<img width="788" alt="image" src="https://github.com/OpenSalamander/salamander/assets/230851/7257d23d-cb74-4b21-bd7e-8e131e519935">

**Comment by 0xeb** [comment]
btw, reviewing some changes before PR, GPT-4 is not that better:

![image](https://github.com/OpenSalamander/salamander/assets/230851/2d1ea538-d0ef-40ce-a3c7-c9decbe2d2ba)

I think "Harder" is better "Hard exit" instead.

**Reply by 0xeb** [reply]
GPT4
![image](https://github.com/OpenSalamander/salamander/assets/230851/9070c0cd-eb03-40ae-8530-008e6a39351f)
vs GPT 3.5:
![image](https://github.com/OpenSalamander/salamander/assets/230851/669d7230-ac67-436b-a0cb-c45f5f97f792)

GPT3.5 is better here.

**Comment by 0xeb** [comment]
This translation work will be tedious but I believe we should all work together. Since you know Czech, the best approach is if you start editing.

But we need to keep the Git commit authorship for the best open source spirit.

I am proposing to give you owners commit access to my fork.

When I make a PR from my fork/branch, go ahead, switch to my branch and edit directly, then push to my branch again.
This will reflect in the PR as well.

That will move things fast.

If there are too many automatic translation errors, I can re-generate stuff again. There will be some things that are hard to automate. Like some spacing missing here and there (but not always). Sometimes comments seem disconnected because we translate one comment at a time instead of a group of comments (where then the whole comment has a 'context'), etc.

Any comments?

**Comment by janrysavy** [comment]
I asked other (Czech) users if they would join the translation project. It would speed up the project considerably.

https://forum.altap.cz/viewtopic.php?p=73678#p73678

**Reply by janrysavy** [reply]
I don't know if we understand each other. We have translated several large files, all done manually, using Chat GPT-4, DeepL, Google Translator. I then did a careful review and corrections on them.

If we were able to feed these translated Czech / English pairs as a default template or dictionary for automatic translation, the final translation would be better. I haven't studied the Chat GPT API if it's ready for this, but I'm sure it would make sense in the overall scope of the translation project.

**Reply by 0xeb** [reply]
Let me clarify, all the source CPP files are translated now and I believe they are 90% quality and need review. So practically speaking, we got the translation tasked knocked out and done, pending review. That's much more sustainable than dragging this work for a few more month.



How do you suggest we proceed w/ the help of the community?

Best way is to see who wants to fork my branch and fix the translations, then we merge everything.

Or as I told you as well, you are welcome to checkout my branch and edit directly, then update the commit.
